src/passes/OptimizeInstructions.cpp

```diff
@@ -2712,6 +2715,32 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  // We can combine `and`, `or`, `xor` operations, e.g.
+  //   (X op Z) ^ (Y op Z)  ==>   (x ^ y) op Z
+  //   (X op Z) | (Y op Z)  ==>   (x | y) op Z
+  //   (X op Z) & (Y op Z)  ==>   (x & y) op Z
+  Expression* combineAnyBitwise(Binary* curr) {
+    using namespace Abstract;
+    using namespace Match;
+    {
+      BinaryOp op;
+      Binary *bx, *by;
+      Expression *x, *y, *z1, *z2;
+      if (matches(curr,
+                  binary(&op,
+                         binary(&bx, any(&x), any(&z1)),
+                         binary(&by, any(&y), any(&z2)))) &&
+          bx->op == by->op && hasAnyBitwise(op) && canReorder(z1, y) &&
+          areConsecutiveInputsEqualAndFoldable(z1, z2) && preserveShift(bx)) {
```

Comment:
This checks "has any bitwise" on the inner operation, and then "preserve shift" on the outside operation. That doesn't match. I'd expect to see us check the inner op has property X and the outer op preserves that same property X.

Reply:
I added extra `bx->type == curr->type`. Will that be enough? Unfortunately, I don't know how to generalize the rule more. So `hasAnyBitwise` is bounds to `preserveShift` for now. They work in tandem by imposing a one-to-one restriction. One for external expression and the other for the inner expression.